### PR TITLE
Refine diversity model

### PR DESCRIPTION
**Pull request by isabella618033**
## Model update and benchmarking result
- Changed the embedding mode from bert_base_uncased to "sentence-transformers/all-mpnet-base-v2"

[fig 1]
SNLI_dataset
- here shows the similarity scores from mpnet and bert_base model respectively on the [SNLI dataset](https://huggingface.co/datasets/snli), which focuses on recognizing textual entailment
- label 0 (green) refers to entailment sentences; label 2 (blue) refers to contradiction;  and label 1 (red) refers to contradiction
- mpnet is outperforming bert_base in classifying the 2 (entailment VS contradiction) classes with threshold at ~0.5
- bert_base is failing at giving us clearity in the difference in the dataset

[fig 2]
c4_200m_dataset
- it shows the benchmark result from mpnet and bert_base model respectively on the [c4 200m dataset](https://huggingface.co/datasets/liweili/c4_200m), which focuses on grammatical error correction (GEC) tasks, so that it can help us see the sensitivity of the model on some slight changes of the sentences. The less sensitive it is the better.
- note that although it seems that the model scoring are more diversed in mpnet, all of the values are larger then the TH we have set in fig 1 (> 0.5), so those should be deemed as relevent to each other, so mpnet is accurate in this case. 

[fig 3]
foundation_validator_run
- it shows the benchmark result from mpnet and bert_base model respectively on a validator run 
- mpnet is giving a more diversified score then bert_base

## Similarity reward conversion
- Changed the scoring from mean difference to 10% quantile, meaning that the diversity reward score would be lowered if your response is the same as 10% of the network. 
This PR here changes the the conversion of similarity to reward
From:
```python
1 - ( mean_similarity / torch.mean( mean_similarity ) )
```
To:
```python
(1 - similarity).quantile(0.1, dim = 1 )
```
For the following reasons
- Taking mean may not be good because if all peers are controlled by the same group, then they can easily separate themselves into 2 groups and response with 2 very different sentences in each group,
- Taking min may not be good because I think we still have to reserve some possibility where a honest server can still have about the same response with the rest of the network.
- So, if we take a quantile at 0.1 (adjustable), then their response can only be repeated with 10% of the network to maximise their diversity reward. So it could be a better approach then mean/min

### CODE REVIEW

**Review by opentaco — APPROVED** [review]
Could you add a motivation statement for the 10% quantile choice in this PR, just to complete the record.
Also, since it seems a negative similarity is possible in practice, that case need to be handled, and it seems like a clamp preventing negative vals may be simplest.

File: openvalidators/reward/diversity.py [hunk 1]
```diff
@@ -37,65 +38,63 @@ def mean_pooling( model_output, attention_mask ):
...
     def name(self) -> str: return RewardModelType.diversity.value
 
     def __init__( self, device: str ):
         super().__init__()
         self.device = device
-        self.tokenizer = AutoTokenizer.from_pretrained( DiversityRewardModel.relevance_model_path )
-        self.model = AutoModel.from_pretrained( DiversityRewardModel.relevance_model_path ).to(self.device)
-
-    def get_embedding( self, message: str ) -> "torch.FloatTensor":
+        self.tokenizer = AutoTokenizer.from_pretrained( DiversityRewardModel.diversity_model_path )
```
Comment (Eugene-hu) [inline openvalidators/reward/diversity.py:56]: What is the sequence length of the model? will there be an issue if we exceed it?

**Review by Eugene-hu — APPROVED** [review]
LGTM

**Comment by steffencruz** [comment]
Thorough and interesting analysis

**Comment by steffencruz** [comment]
# C4 
> help us see the sensitivity of the model on some slight changes of the sentences. The less sensitive it is the better.

Love this idea. We need to be aware that we are assuming that the before/after sentence are semantically equivalent. From looking at the raw dataset, it's not always the true.

image

For example: `"Worthless involved's supporting for the movement."` is not equivalent to `"Get involved and help the movement!"`

This is probably what is causing the long tail in your plot. 

> note that although it seems that the model scoring are more diversed in mpnet, all of the values are larger then the TH we have set in fig 1 (> 0.5), so those should be deemed as relevent to each other, so mpnet is accurate in this case.

Looks like the `mpnet` values are sometimes below 0.5. 

Again, it would be helpful to know how the plotted values were produced.

**Comment by steffencruz** [comment]
> Could you add a motivation statement for the 10% quantile choice in this PR, just to complete the record. Also, since it seems a negative similarity is possible in practice, that case need to be handled, and it seems like a clamp preventing negative vals may be simplest.

Agree with @opentaco . Before/after plots of diversity would be ideal. Hit me up if you need me to jump on any of these tasks :)

**Comment by isabella618033** [comment]
Thank you for asking questions about the experiments :)

> A code snippet of how the plotted values were produced would be helpful here. My guess is you group the dataset by context/premise and then model the completions as the hypotheses?
> 
> ```python
> DiversityRewardModel .get_rewards(prompt=premise, completions=hypotheses)
> ```

So the way that I was using the diversity model would be: 

```python
DiversityRewardModel .get_rewards(prompt="", completions=[ premise, hypotheses ])`
```
 
That is because in the diversity model, what we want to know is the comparisons between completions and it is not related to the prompt.  

> # SNLI
> Assuming the intended labels are equivalent to [snli](https://huggingface.co/datasets/snli): 0 - entailment (logically **coherent** with the prompt) 1 - neutral (**somewhat irrelevant** to the prompt) 2- contradiction (logically **incoherent** with the prompt)
> 

So the correspondence usage of the dataset would be the following
- 0 - entailment ( completions are logically coherent, thus related to each other or mean about the same thing, so result in less diversity)
- 1 - neutral ( completions are somewhat irrelevant to each other, so more diversity)
- 2- contradiction ( completions are logically incoherent with each other or not even related to each other, thus the most diversity)

> image
> 
> What are our expectations about the diversity scores on this dataset, and how do the results prove or disprove that the sentence embedding model replacement is superior?

So from this figure
image 

- You can see that bert_base fails to identify all of these relationships (0/1/2), and deem all of these to have the same level of correlation to each other; 
- while mpnet can success in this task by scoring 0 higher and 2 lower on average.    

> I'm struggling to see the link between entailment and diversity. Seems more like a test of relevance?

So it really depends on what kind of task are we expecting from the models. 
- relevance model was designed for filtering out some trivial completions that is not related to the prompt
- the diversity model here actually main aims to bring more diversity completion to the network by tackling the look up attack (at least for the first step). A more advanced look up attack would slightly augmenting the original retrieved completion, so we have to bring more robustness to the diversity/similarity/entailment by benchmarking with the entailment dataset.

**Comment by isabella618033** [comment]
> # C4
> > help us see the sensitivity of the model on some slight changes of the sentences. The less sensitive it is the better.
> 
> Love this idea. We need to be aware that we are assuming that the before/after sentence are semantically equivalent. From looking at the raw dataset, it's not always the true.
> 
> image
> 
> For example: `"Worthless involved's supporting for the movement."` is not equivalent to `"Get involved and help the movement!"`
> 
> This is probably what is causing the long tail in your plot.
> 
> > note that although it seems that the model scoring are more diversed in mpnet, all of the values are larger then the TH we have set in fig 1 (> 0.5), so those should be deemed as relevent to each other, so mpnet is accurate in this case.
> 
> Looks like the `mpnet` values are sometimes below 0.5.
>
Yes, I agree. This dataset is not the best in our use case, with somewhat changed semantic in the midst of changing the sentence structure, which causes the long tail and having the `mpnet` values goes below 0.5. 

So in this case, `mpnet` would be better then `bert_base` if we find out that there is a large semantic change in the dataset. And `bert_base` would be better then `mpnet` otherwise. 
   
Moreover, I hope that this test on the C4 dataset can help us get a sense of the meaning of the `mpnet` score or `bert_base` score at different range.
Let me send you a resulted CSV here for your interest. I think it would be better at illustrating the difference in score / completions. 

> Again, it would be helpful to know how the plotted values were produced. Not clear how this addresses the questions of diversity..

The values was produced by the same get rewards function without the quantile stuff.
```python
DiversityRewardModel .get_rewards(prompt="", completions=[ premise, hypotheses ])`

    def get_rewards( self, prompt: str, completions: List[str], name: str ) -> torch.FloatTensor:

        # Get embeddings for all completions.
        embeddings = self.get_embeddings( completions )

        # Calculate the pairwise cosine similarity.
        similarity = pairwise_cosine_similarity( embeddings, embeddings )

        # Return all
        return similarity

``` 
(just copying from the previous comment)
- the diversity model here actually mainly aims to bring more diversified completion to the network by tackling the look up attack (at least for the first step). A more advanced look up attack would slightly augmenting the original retrieved completion, so we have to bring robustness to the diversity/similarity/entailment model by benchmarking them with the entailment dataset.

**Comment by isabella618033** [comment]
> Could you add a motivation statement for the 10% quantile choice in this PR, just to complete the record. 

So this PR here changes the the conversion of similarity to reward
From:
 ```python 
1 - ( mean_similarity / torch.mean( mean_similarity ) )
```
To:
```python
(1 - similarity).quantile(0.1, dim = 1 )
```
 - Taking mean may not be good because if all peers are controlled by the same group, then they can easily separate themselves into 2 groups and response with 2 very different sentences in each group,
- Taking min may not be good because I think we still have to reserve some possibility where a honest server can still have about the same response with the rest of the network.
 - If we take a quantile at 0.1 (adjustable), then their response can only be repeated with 10% of the network to maximise their diversity reward. 

> Also, since it seems a negative similarity is possible in practice, that case need to be handled, and it seems like a clamp preventing negative vals may be simplest.

Sure! will make the change accordingly.

File: openvalidators/reward/diversity.py (hunk 1 above)
Comment (isabella618033) [reply openvalidators/reward/diversity.py:56]: Nice catch! max_len 512, let me truncate it.

**Review by steffencruz — APPROVED** [review]
LGTM
